src/internal/operators/tap.ts

```diff
@@ -18,131 +19,165 @@ export function tap<T>(observer: PartialObserver<T>): MonoTypeOperatorFunction<T
+ * using `tap`.
+ *
+ * ```ts
+ * import { of } from 'rxjs':
+ * import { tap } from 'rxjs/operators';
+ *
+ * const source = of(1, 2, 3, 4, 5)
+ *
+ * source.pipe(
+ *  tap(n => {
+ *    if (n > 3) {
+ *      throw new TypeError(`Value ${n} is greater than 3`)
+ *    }
+ *  })
+ * )
+ * .subscribe(console.log);
+ * ```
+ *
+ * ## Example
+ * We want to know when an observable completes before moving on to the next observable. The system
+ * below will emit a random series of `"X"` characters from 3 different observables in sequence. The
+ * only way we know when one observable completes and moves to the next one, in this case, is because
+ * we have added a `tap` with the side-effect of logging to console.
+ *
+ * ```ts
+ * import { of, interval } from 'rxjs';
+ * import { tap, concatMap, take } from 'rxjs';
+ *
+ *
+ * of(1, 2, 3).pipe(
+ *  concatMap(n => interval.pipe(
+ *    take(Math.round(Math.random() * 10)),
+ *    map(() => 'X'),
+ *    tap({
+ *      complete: () => console.log(`Done with ${n}`)
+ *    })
+ *  ))
+ * )
+ * .subscribe(console.log);
  * ```
  *
- * @see {@link map}
+ * @see {@link finalize}
  * @see {@link Observable#subscribe}
  *
- * @param {Observer|function} [nextOrObserver] A normal Observer object or a
- * callback for `next`.
- * @param {function} [error] Callback for errors in the source.
- * @param {function} [complete] Callback for the completion of the source.
- * @return {Observable} An Observable identical to the source, but runs the
- * specified Observer or callback(s) for each item.
- * @name tap
+ * @param nextOrObserver A next handler or partial observer
+ * @param error An error handler
+ * @param complete A completion handler
  */
-export function tap<T>(nextOrObserver?: PartialObserver<T> | ((x: T) => void) | null,
-                       error?: ((e: any) => void) | null,
-                       complete?: (() => void) | null): MonoTypeOperatorFunction<T> {
-  return function tapOperatorFunction(source: Observable<T>): Observable<T> {
-    return lift(source, new TapOperator(nextOrObserver, error, complete));
-  };
-}
+export function tap<T>(
+  nextOrObserver?: PartialObserver<T> | ((x: T) => void) | null,
+  error?: ((e: any) => void) | null,
+  complete?: (() => void) | null
+): MonoTypeOperatorFunction<T> {
+  return (source: Observable<T>): Observable<T> => {
+    return lift(source, function (this: Subscriber<T>, source: Observable<T>) {
+      const subscriber = this;
+      let onNext: (value: T) => void;
+      /**
+       * A helper to ensure that errors thrown in handlers get
+       * caught and sent do the consumer as an error notification.
+       */
+      const wrap = (fn: any) => (arg?: any) => {
+        try {
+          fn(arg);
+        } catch (err) {
+          subscriber.error(err);
+        }
+      };
+      let onError = wrap(error ?? noop);
+      let onComplete = wrap(complete ?? noop);
```

Comment:
question: Is there a reason the calls to `wrap` for `error` and `complete` are not inside the `if (!nextOrObserver || typeof nextOrObserver === 'function') { ... }` block? AFAICT, they will pointlessly wrap `noop`s when an observer is passed.

Reply:
Probably playing too much code golf.

Reply:
Ironically, correcting these goofs resulted in an even smaller operator. Added to the PR. I'll just merge after this passes CI.

Now: 549B

![tap-after](https://user-images.githubusercontent.com/1540597/92931290-dd6c4f00-f408-11ea-8309-e6f243552cf0.png)